### PR TITLE
[idea] Use the Observer Pattern instead of the Decorator Pattern

### PR DESCRIPTION
**Pull request by danthegoodman**
Relevant Link: [Wikipedia - Observer Pattern](https://en.wikipedia.org/wiki/Observer_pattern)

Following up on my comments in my previous PR #3:
> I disagree with your decision to use the decorator pattern on the canvas.

Talk is cheap. My comment has set heavy in my mind for the last few days. I didn't feel comfortable just making claims and walking away, so I wrestled with your code a bit and I think I came a pattern that feels a little more appropriate for the project. 

Before I go any further, please do not feel obligated to take this PR or to even use it in the slightest. I've been looking for things to play with for [Hacktoberfest](https://hacktoberfest.digitalocean.com), and following up my statement seemed like and was a fun challenge. If you like the ideas here, take them and run with them in whatever way you want! You can reimplement it yourself or you can merge the PR or you can completely dismiss my input. I won't suffer any hard feelings for any decision you choose.

Back to the idea: I think two things unsettled me most about the decorator pattern here. First, the decorator was used to merge the Grid class and ASCIICanvas class. It seems to me that if the two classes are implemented separately, they should be used separately. So, in this PR, I did so.

Second, and this is the bigger one: The decorator pattern was mostly used as a means to propagate events to different components. In all of the mouseUp, keyDown and similar event methods, the decorator was only appending logic after the base method was called. There wasn't any decorating going on, just additional logic. I think this would be a better use case for the Observer Pattern instead. I did this through the `CanvasEvents` class, which is available on ASCIICanvas. This allows the other canvases and decorators to be notified when things happen like KeyPresses, MouseUps, and even custom events like Resize and CellDown.

That said, I did find one place where the decorator pattern was used to fully decorate behavior. The PointerDecorator truely decorated the `hasChanged` behavior. I think applying the decorator just for this is a little overkill, so I refactored how the animation function is used to instead use both the ASCII Canvas and the PointerDecorator.

I _did_ leave the full blown proxy decoration in one place: the StylableCanvas. I think the drawLine function is another example of something that could be an event, I just didn't take the time to actually turn it into another event. I even thought about ripping out the StylableCanvas all together, because it's really only consists of one extra long function, and that function could be more of a "Stylable Util" than trying to be a decorator.

In conclusion, I hope my opinions don't come off too harshly. In programming there's always multiple solutions to a problem, and someone else will always have different opinions from you. While i've been working on this branch, I titled it the `really_opinionated_refactor`. I'm opening up the PR as an idea for you, not as something I expect you to use. I hope this is somewhat helpful for you, or that you at least learn or rediscover something as a result of this. Cheers. 🍻

### CODE REVIEW

**Comment by andresoviedo** [comment]
Thanks @danthegoodman Thanks for your valuable and detailed feedback. I will check your idea asap.

As for now, I partly agree with you.  However, let me explain why I used the Decorator Pattern.

The Decorator Patterns is used to add behaviour to the base "HTMLCanvasElement". 
The idea on using this pattern was to add new features (new methods) or responsabilities to this HTML canvas in a decoupled way.

There are the following Decorators for the ASCIICanvas class:
* MovableCanvas : add scrolling capabilities
* ZoomableCanvas : add zoom capabilities
* DrawableCanvas : add draw lines capabilites
* StylableCanvas : add UTF-8 ASCII capabilities
* PointerDecorator : add pointer to screen
* WritableCanvas : add writing capabilities

From this list, every decorator has it's own responsabilities and are independent from each other(*).  Every decorator is really decorating base ASCIICanvas class, by adding new methods to it's prototype. The javascript way used to decorate class was to extend prototype by using javascript Object.create() (although at some time I used the proxy pattern with delegation to support decoration).  So, I think I am indeed decorating ASCIICanvas.

* It's true that I used decoration to merge Grid and ASCIICanvas, but was that just for convenience. Probably an Aggregation pattern would define precisely this association.
* It's true that decorators like WritableCanvas or StylableCanvas depends on DrawableCanvas and that may be a mix of Decoration and Inheritance.  Probably a Wrapper, Delegation or Inheritance would define better the dependency.
* It is true that I have mixed Decoration, Observation and Delegation in Decorator class. I know that could be improved. Probably a split of these patterns would make it clearer.

Thank you again for your comments. I am sure it will help to find a more clean and scalable solution.

Kind regards
